rclcpp_components/src/component_manager.cpp

```diff
@@ -0,0 +1,203 @@
+    if (parts.size() != 2) {
+      RCLCPP_ERROR(get_logger(), "Invalid resource entry");
+      response->error_message = "Invalid resource entry";
+      response->success = false;
+      return;
+    }
+    // match plugin name with the same rmw suffix as this executable
+    if (parts[0] != request->plugin_name) {
+      continue;
+    }
+
+    std::string class_name = parts[0];
+    std::string fq_class_name = "rclcpp_components::NodeFactoryTemplate<" + class_name + ">";
+
+    // load node plugin
+    std::string library_path = parts[1];
+    if (!rcpputils::fs::path(library_path).is_absolute()) {
+      library_path = base_path + "/" + library_path;
+    }
+    class_loader::ClassLoader * loader;
+
+    if (loaders_.find(library_path) != loaders_.end()) {
+      loader = loaders_[library_path].get();
+    } else {
+      RCLCPP_INFO(get_logger(), "Load library %s", library_path.c_str());
+      try {
+        loaders_[library_path] = std::make_unique<class_loader::ClassLoader>(library_path);
+        loader = loaders_[library_path].get();
+      } catch (const std::exception & ex) {
+        RCLCPP_ERROR(get_logger(), "Failed to load library: %s", ex.what());
+        response->error_message = "Failed to load library";
+        response->success = false;
+        return;
+      } catch (...) {
+        RCLCPP_ERROR(get_logger(), "Failed to load library");
+        response->error_message = "Failed to load library";
+        response->success = false;
+        return;
+      }
+    }
+    auto classes = loader->getAvailableClasses<rclcpp_components::NodeFactory>();
+    for (const auto & clazz : classes) {
+      RCLCPP_INFO(get_logger(), "Found class %s", class_name.c_str());
+      if (clazz == class_name || clazz == fq_class_name) {
+        RCLCPP_INFO(get_logger(), "Instantiate class %s", class_name.c_str());
+        auto node_factory = loader->createInstance<rclcpp_components::NodeFactory>(clazz);
+
+        std::vector<rclcpp::Parameter> parameters;
+        for (const auto & p : request->parameters) {
+          parameters.push_back(rclcpp::Parameter::from_parameter_msg(p));
+        }
+
+        std::vector<std::string> remap_rules {request->remap_rules};
+
+        if (!request->node_name.empty()) {
+          remap_rules.push_back("__node:=" + request->node_name);
+        }
+
+        if (!request->node_namespace.empty()) {
+          remap_rules.push_back("__ns:=" + request->node_namespace);
+        }
+
+        auto options = rclcpp::NodeOptions()
+          .initial_parameters(parameters)
+          .arguments(remap_rules);
+
+        auto node_id = unique_id++;
+
+        node_wrappers_[node_id] = node_factory->create_node_instance(options);
+
+        auto node = node_wrappers_[node_id].get_node_base_interface();
+        if (auto exec = executor_.lock()) {
+          exec->add_node(node, true);
+        }
+        response->full_node_name = node->get_fully_qualified_name();
```

Comment:
The [design pr](https://github.com/ros2/design/pull/206/files#diff-df7db447d3b09c67c5487a923f56a445R527) required this to cancel loading the node if there is a name conflict, but now I'm thinking that sentence should be removed in favor of referencing a design doc that satisfies ros2/design#187. I don't think anything here needs to change.